umbraco/Umbraco-CMS
# New Backoffice: Log viewer controller

## Details
- Created a new service for the LogViewer that gathers implementations from `ILogViewer`, `ILogViewerConfig` and `ILogLevelLoader`;
- New endpoints for the Log Viewer:
  - ~GET `/log-viewer/error/count` - gets the number of errors from the logs;~ _(removed in commit c470f6cda12333e72706be0906c71966a59a4d59 since the number can be derived from `/log-viewer/level-count`, summing Error + Fatal)_;
  - GET `/log-viewer/level` - gets all loggers' levels in a paginated list;
  - GET `/log-viewer/level-count` - gets the count for each log level from the logs;
  - GET `/log-viewer/log` - gets all logs in a paginated list;
    ~- A temp solution for representing enums as strings for solving an issue with Swagger has been added in commit c6e4ac35f7b228da879c66d2d75380a9159b7f7f;~ If you want to test with selecting the enums types, it has to be in Postman;
    - Note ❗ there is currently a polymorphism issue that we will work on in the future, mentioned in [comment](https://github.com/umbraco/Umbraco-CMS/blob/v13/new-backoffice/log-viewer-controller/src/Umbraco.Infrastructure/Services/Implement/LogViewerService.cs#L219-L220). The problem we see is in the properties of some log entries with Error log level, like:
 > Before:
```json
{
  . . .
  "items": [
    {
      . . .
      "Level": "Error",
	  . . .
      "Properties": {
        "EventId": {
          "TypeTag": null,
          "Properties": [
            {
              "Name": "Id",
              "Value": {
                "Value": 1
              }
            },
            {
              "Name": "Name",
              "Value": {
                "Value": "UnhandledException"
              }
            }
          ]
        }
      }
    }
  ]
}
```
 > Now (_The nested `Value` object is empty_):
```json
{
  . . .
  "items": [
    {
      . . .
      "level": "Error",
	  . . .
      "properties": [
        {
          "name": "EventId",
          "value": "{\"TypeTag\":null,\"Properties\":[{\"Name\":\"Id\",\"Value\":{}},{\"Name\":\"Name\",\"Value\":{}}]}"
        }
      ]
    }
  ]
}
```
- 
  - GET `/log-viewer/message-template` - gets all log message templates (_used to determine common log patterns or errors_) in a paginated list;
  - GET `/log-viewer/saved-search` - gets all saved log searches in a paginated list;
  - POST `/log-viewer/saved-search` - creates a saved log search;
  - GET `/log-viewer/saved-search/{name}` - gets a saved log search by name;
  - DELETE `/log-viewer/saved-search/{name}` - deletes a saved log search with a given name;
  - GET `/log-viewer/validate-logs-size` - gets a response on whether you are able to view logs for a specified date range (<em><code>true</code> when the requested log files are less than 1GB in size</em>)
- Cherry-pick  #13634 and #13647;
- Introducing `GetLogsAsPagedModel()`+ update for it in a _suppressions.xml_ file;

## Test
- You can use Swagger to test the endpoints - visit `localhost/umbraco/swagger`;
~- `/log-viewer/error/count` - _GET_~
  ~- Compare the number of errors for a given time period, like `startDate=2023-01-09` and `endDate=2023-01-10`, with the one shown on the _Log Viewer_ dashboard.~
- `/log-viewer/level` - _GET_
  - Execute with the default _skip_ and _take_ values or modify them as you like;
  - Compare with the response from the backoffice when going to the _Log Viewer_ - _the response from `GetLogLevels()`_.
- `/log-viewer/level-count` - _GET_
  - Compare with the response from the backoffice when going to the _Log Viewer_ - _the response from `GetLogLevelCounts()`_.
- `/log-viewer/log` - _GET_
  - Compare with the response (`items[]`) from the backoffice when going to the _Log Viewer_ > _Saved Searches (All Logs)_;
    - You can filter the logs by log level, like `logLevel=Error&logLevel=Fatal`;
      - _Note:_ in Swagger you can select several levels by holding _Ctrl_.
    - Or add a filter expression, like `Not(@Level='Verbose') and Not(@Level='Debug')` (_check the default Saved Searches for other examples_);
    - etc.
- `/log-viewer/message-template` - _GET_
  - Compare with the response from the backoffice when going to the _Log Viewer_ - _the response from `GetMessageTemplates()`_;
  - They can be seen in _Log Viewer_ > _Common Log Messages_.
- `/log-viewer/saved-search` - _GET_
  - Compare with the response from the backoffice when going to the _Log Viewer_ - _the response from `GetSavedSearches()`_;
  - They can be seen in _Log Viewer_ > _Saved Searches_.
- `/log-viewer/saved-search` - _POST_
  - Create a new saved search by providing a name and a query value, like:
```json
{
  "name": "Find all logs where the Level is NOT Information and NOT Warning",
  "query": "Not(@Level='Information') and Not(@Level='Warning')"
}
```
-
  - Verify that the response headers contain a **Location** response header, indicating the URL to get the details about this item;
  - Try creating a saved search with a name that already exists - the response should be with status 400 and title _"Duplicate log search name"_.
- `/log-viewer/saved-search/{name}` - _GET_
  - _New endpoint_: get all saved searches from GET `/log-viewer/saved-search` and find a name of a saved search;
  - Check if the response contains the right query.
- `/log-viewer/saved-search/{name}` - _DELETE_
  - Pass the name of a saved search;
  - Verify that if the item exists the response is OK, otherwise it will be Not Found.
- `/log-viewer/validate-logs-size` - _GET_
  - Compare with the response from the backoffice when going to the _Log Viewer_ - _the response from `GetCanViewLogs()`_.
